### PR TITLE
fix: load ImageService config

### PR DESCRIPTION
**Pull request by HaokunX-intel**
Load configs at the agent launch stage instead of
before the image pulling.

Signed-off-by: Xing, Haokun <haokun.xin@intel.com>

### CODE REVIEW

**Comment by hairongchen** [comment]
cosign now is implemented as a feature of image-rs, so need to add to the feature list as bellow

> features = ["occlum_feature", "**cosign**"]

**Comment by HaokunX-intel** [comment]
> cosign now is implemented as a feature of image-rs, so need to add to the feature list as bellow
> 
> > features = ["occlum_feature", "**cosign**"]

This PR loads config early and enables signature verification, regardless of client types ( `grpc-based client` or `native client`) or signature types (`single sign` or `cosign`). 

And this PR depends on `tag = 0.2.0` of `image-rs`, which has no `cosign` feature. We may propose another issue to enable `cosign`.

File: src/enclave-agent/src/config.rs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+            path: String,
+            result_key_provider: Option<String>,
+            result_security_validate: Option<bool>,
+        }
+
+        let config_cases: Vec<ConfigCase> = vec![
+            ConfigCase {
+                path: "test_data/decrypt_config/decrypt_config_normal.conf".to_string(),
+                result_key_provider: Some(
+                    "provider:attestation-agent:eaa_kbc::127.0.0.1:1234".to_string(),
```
Comment (Xynnn007) [inline src/enclave-agent/src/config.rs:75]: Could this constant `"provider:attestation-agent:eaa_kbc::127.0.0.1:1234"` be defined as a `const` in this test mod? Same fow below

File: src/enclave-agent/src/config.rs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        }
+    }
+}
+"#;
+
+#[derive(Serialize, Deserialize, Default)]
+pub struct DecryptConfig {
+    pub key_provider: Option<String>,
+    #[serde(default = "default_security_validate")]
+    pub security_validate: Option<bool>,
```
Comment (Xynnn007) [inline src/enclave-agent/src/config.rs:19]: Another flag is the switch for `auth`. related link for `ImageConfig` https://github.com/confidential-containers/image-rs/blob/main/src/config.rs#L29. I'm afraid we also need to add one flag for that?

**Review by Xynnn007 — COMMENTED** [review]
Overall, LGTM.

Also, I think a configuration file/struct is good for image-rs. Maybe we can consider bring this mechinism directly into image-rs in the future to support
* more flexible configurations for image-rs, like auth
* counterpart for kata-cc

What is your opinion upon this? @arronwy

File: src/enclave-agent/src/config.rs (hunk 2 above)
Comment (HaokunX-intel) [reply src/enclave-agent/src/config.rs:19]: The flag for `auth` has been added.

Comment (Xynnn007) [reply src/enclave-agent/src/config.rs:19]: Great, sorry for neglecting

**Review by Xynnn007 — APPROVED** [review]
Thanks @HaokunX-intel  LGTM

File: src/enclave-agent/src/services/images.rs [hunk 3]
```diff
@@ -136,42 +100,45 @@ mod test {
 
     #[test]
     fn test_get_container_id() {
-        struct ParseCase{
+        struct ParseCase {
```
Comment (mythi) [inline src/enclave-agent/src/services/images.rs:103]: let's move formatting fixes to another pr

File: src/enclave-agent/src/config.rs [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        std::env::set_var("OCICRYPT_KEYPROVIDER_CONFIG", self.config_path.clone());
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn test_load_decrypt_config() {
```
Comment (mythi) [inline src/enclave-agent/src/config.rs:70]: I think this is not testing what we are interested in:  errors are correctly returned when file read or json parsing fails based on inputs provided.

File: src/enclave-agent/Cargo.toml [hunk 5]
```diff
@@ -20,14 +20,15 @@ ctrlc = { version = "3.0", features = ["termination"] }
 tokio = { version = "1.14.0", features = ["full"] }
 async-trait = "0.1.42"
 
-image-rs = { git = "https://github.com/confidential-containers/image-rs", tag = "v0.2.0", features = ["occlum_feature"], default-features = false }
+image-rs = { git = "https://github.com/confidential-containers/image-rs", "rev" = "78a5114", features = ["occlum_feature", "cosign"], default-features = false }
```
Comment (mythi) [inline src/enclave-agent/Cargo.toml:23]: is it needed in this PR?

**Review by mythi — CHANGES_REQUESTED** [review]

File: src/enclave-agent/Cargo.toml (hunk 5 above)
Comment (HaokunX-intel) [reply src/enclave-agent/Cargo.toml:23]: > Another flag is the switch for `auth`. related link for `ImageConfig` https://github.com/confidential-containers/image-rs/blob/main/src/config.rs#L29. I'm afraid we also need to add one flag for that?

The `auth` mentioned by @Xynnn007 is introduced after `tag = "v0.2.0"`.

File: src/enclave-agent/src/config.rs (hunk 4 above)
Comment (HaokunX-intel) [reply src/enclave-agent/src/config.rs:70]: The rust standard create provides `file read` APIs, it is out of the responsibilities to cover all file read failures.

The crate `serde-json` provides `json parsing` APIs. It will return different error info once reading invalid json files, it is hard to emulate all cases.

I think we at most add two cases, including reading a not existing file, reading a json files with invalid key, to represent file read failure and json parse failure.

File: src/enclave-agent/src/services/images.rs (hunk 3 above)
Comment (HaokunX-intel) [reply src/enclave-agent/src/services/images.rs:103]: I think these formatting fixes are too few to deserve to propose another PR.

Comment (mythi) [reply src/enclave-agent/src/services/images.rs:103]: a separate commit in this PR is also fine.

File: src/enclave-agent/src/config.rs (hunk 4 above)
Comment (mythi) [reply src/enclave-agent/src/config.rs:70]: > I think we at most add two cases, including reading a not existing file, reading a json files with invalid key, to represent file read failure and json parse failure.

makes sense

File: src/enclave-agent/Cargo.toml (hunk 5 above)
Comment (mythi) [reply src/enclave-agent/Cargo.toml:23]: these are not related to the fix in this PR and in fact now it's blocking getting this PR merged. I'd suggest we drop `image-rs` updates in this PR so we get it merged.

File: src/enclave-agent/src/services/images.rs (hunk 3 above)
Comment (HaokunX-intel) [reply src/enclave-agent/src/services/images.rs:103]: It sounds good.

File: src/enclave-agent/Cargo.toml (hunk 5 above)
Comment (Xynnn007) [reply src/enclave-agent/Cargo.toml:23]: Agreed. We can open an issue to trace registry authentication feature and impl in another PR.

Comment (HaokunX-intel) [reply src/enclave-agent/Cargo.toml:23]: Sure, I'll drop the `auth` this PR.

File: src/enclave-agent/src/services/images.rs (hunk 3 above)
Comment (mythi) [reply src/enclave-agent/src/services/images.rs:102]: all of these `test_get_container_id`  formatting changes can go into 5215746a32878641ce213086de3c2fce5af050bc

File: src/enclave-agent/src/config.rs [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+            } else {
+                let c = result.unwrap();
+                assert_eq!(c.key_provider, case.result_key_provider);
+                assert_eq!(c.security_validate, case.result_security_validate);
+            }
+        }
+    }
+
+    #[test]
+    fn test_keyprovider_config_default() {
```
Comment (mythi) [inline src/enclave-agent/src/config.rs:125]: It fails because the test case is not allowed to write to `DEFAULT_KEYPROVIDER_CONFIG_PATH`:

```
---- config::test::test_keyprovider_config_default stdout ----
thread 'config::test::test_keyprovider_config_default' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 13, kind: PermissionDenied, message: "Permission denied" }', src/config.rs:126:54
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

File: src/enclave-agent/src/main.rs [hunk 7]
```diff
@@ -45,7 +48,16 @@ async fn main() -> Result<()> {
         TCP_SOCK_ADDR
     };
 
-    let image_service = Box::new(ImageService::new()) as Box<dyn image_ttrpc::Image + Send + Sync>;
+    let dc = if let Some(file_path) = matches.value_of("decrypt-config") {
+        DecryptConfig::load_from_file(&file_path.to_string())?
```
Comment (mythi) [inline src/enclave-agent/src/main.rs:52]: this path should handle errors better

File: src/enclave-agent/src/config.rs (hunk 6 above)
Comment (HaokunX-intel) [reply src/enclave-agent/src/config.rs:125]: The test case is passed in our env. The agent should run in the root privilege to write to `/etc`. If we want to place the config file in different place, please propose another issue to track the concern.

File: src/enclave-agent/src/services/images.rs (hunk 3 above)
Comment (HaokunX-intel) [reply src/enclave-agent/src/services/images.rs:102]: Actually, this PR affects almost all files.  It is hard and unbeneficial to separate new codes and format fixing into two independent parts completely. Therefore the last commit contains a part of formatting changes, moving all of them seems not necessary.

File: src/enclave-agent/src/config.rs (hunk 6 above)
Comment (mythi) [reply src/enclave-agent/src/config.rs:128]: > The agent should run in the root privilege to write to `/etc`

this is not about running the agent but running the running tests. `cargo test` should pass without root/sudo privileges.

Comment (HaokunX-intel) [reply src/enclave-agent/src/config.rs:128]: If you insist it, please change the [enclave-agent.yaml](https://github.com/confidential-containers/enclave-cc/blob/572a5715bc4941c56e0b4dc0b4dacd3e5a9c4c42/tools/packaging/build/agent-enclave-bundle/enclave-agent.yaml#L25) , copy the ` /run/enclave-agent/decrypt_config.conf` to other directory instead of `/etc`.

File: src/enclave-agent/src/config.rs [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+        let path = DEFAULT_KEYPROVIDER_CONFIG_PATH.to_string();
+        let mut kc = Self {
+            config_path: path.clone(),
+            is_exist: Path::new(&path).exists(),
+        };
+
+        if !kc.is_exist {
+            File::create(Path::new(&path))?.write_all(DEFAULT_KEYPROVIDER_CONFIG.as_bytes())?;
+            kc.is_exist = true;
+        }
```
[the remark this answers is not in the document]
Comment (mythi) [reply src/enclave-agent/src/config.rs:53]: I'm not sure this answers my question. There's the option of giving an error if there's no file pre-installed.

File: src/enclave-agent/src/services/images.rs (hunk 3 above)
Comment (mythi) [reply src/enclave-agent/src/services/images.rs:102]: it's easy to pick the lines to you want using `git add -p ...`

File: src/enclave-agent/src/main.rs (hunk 7 above)
Comment (HaokunX-intel) [reply src/enclave-agent/src/main.rs:52]: Sure. The error will be handled in the `load_from_file`, this path will raise the error if failed.

File: src/enclave-agent/src/services/images.rs (hunk 3 above)
Comment (HaokunX-intel) [reply src/enclave-agent/src/services/images.rs:102]: Sure.

File: src/enclave-agent/src/config.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (HaokunX-intel) [reply src/enclave-agent/src/config.rs:53]: > I'm not sure this answers my question. There's the option of giving an error if there's no file pre-installed.

That is a good idea. But, nowadays we can not customize the config for `ocicrypt-rs`. And we do not provide a default config in the [packaging](https://github.com/confidential-containers/enclave-cc/tree/main/tools/packaging/build).  If the writeback is removed, the agent can not pull and decrypt any encrypted image.

If we want to remove the writeback and pull the encrypted images, we need to add a default config file in the packaging. But if we plan to provide an option to customize the ocicrypt-rs config in the future, the writeback is acceptable.

[the remark this answers is not in the document]
Comment (mythi) [reply src/enclave-agent/src/config.rs:53]: We do provide a default config: https://github.com/confidential-containers/enclave-cc/blob/572a5715bc4941c56e0b4dc0b4dacd3e5a9c4c42/tools/packaging/build/agent-enclave-bundle/enclave-agent.yaml#L25-L28

#62 is also experimenting with the idea of moving the default away from the Occlum instance. Check the plans how this is going to be used in the operator from @hairongchen

[the remark this answers is not in the document]
Comment (mythi) [reply src/enclave-agent/src/config.rs:53]: Yeah, sorry I mixed the two. We can install the keyprovider config too.

[the remark this answers is not in the document]
Comment (HaokunX-intel) [reply src/enclave-agent/src/config.rs:53]: If the keyprovider config is installed by the operator, we can abandon the writeback.

And a new launch option for loading keyprovider config can be provided.

[the remark this answers is not in the document]
Comment (HaokunX-intel) [reply src/enclave-agent/src/config.rs:53]: In this PR, we can install the default keyprovider config in the agent bundle. And #62 supports customized agent config and keyprovider config.

File: src/enclave-agent/src/config.rs (hunk 6 above)
Comment (HaokunX-intel) [reply src/enclave-agent/src/config.rs]: The keyprovider config will not be writeback, the unit test can be run without root privilege now.

File: src/enclave-agent/src/services/images.rs [hunk 9]
```diff
@@ -173,7 +134,11 @@ mod test {
             },
         ];
 
-        let is = ImageService::new();
+        let dc = DecryptConfig::load_from_file(
+            &"test_data/decrypt_config/decrypt_config_normal.conf".to_string(),
```
Comment (mythi) [inline src/enclave-agent/src/services/images.rs:138]: This should be the user provided `decrypt-config` param instead, right?

Comment (HaokunX-intel) [reply src/enclave-agent/src/services/images.rs:138]: Yeah, the user provides the file and specify the path the file when the agent is launched.

**Review by mythi — APPROVED** [review]
